### PR TITLE
Convert vhost repo as a workspace

### PR DESCRIPTION
**Pull request by vireshk**
The workspace will contain vhost and vhost-user-backend crates for now and will get vhost-user-vmm later on.

### CODE REVIEW

**Comment by vireshk** [comment]
I am adding this just for testing right now (as I am getting failures locally), avoid reviews for now please. Thanks.

**Comment by vireshk** [comment]
@andreeaflorescu @slp @eryugey @jiangliu @sboeuf @stefano-garzarella 

This is review ready now. To capture the enitre history of vhost-user-backend, I have performed merge as:

git merge --allow-unrelated-histories vhost-user-backend/main

and so the pull request shows all the existing patches, and errors out in format of one of the ealier patches.

Thanks.

**Comment by andreeaflorescu** [comment]
> @andreeaflorescu @slp @eryugey @jiangliu @sboeuf @stefano-garzarella
> 
> This is review ready now. To capture the enitre history of vhost-user-backend, I have performed merge as:
> 
> git merge --allow-unrelated-histories vhost-user-backend/main
> 
> and so the pull request shows all the existing patches, and errors out in format of one of the ealier patches.
> 
> Thanks.

We can use the administrator permissions to merge this with the commit format failures, that should not be a problem.

**Comment by stefano-garzarella** [comment]
I guess it makes sense to turn this into a workspace, but how do we handle the tags for releases?

Now we only have vhost tags, should we add vhost-user-backend tags as well?

Should we use a prefix? (e.g. vhost-user-backend-v0.7.0).
Or should we align the two versions and use a single tag?

**Comment by andreeaflorescu** [comment]
> I guess it makes sense to turn this into a workspace, but how do we handle the tags for releases?
> 
> Now we only have vhost tags, should we add vhost-user-backend tags as well?
> 
> Should we use a prefix? (e.g. vhost-user-backend-v0.7.0). Or should we align the two versions and use a single tag?

This is something that we had to do for vm-virtio as well. Initially the crates were in separate repositories. For releases we're just using the name of the crate in the tags. There are some other things that are different with workspaces when doing releases, and they're documented here: https://github.com/rust-vmm/community/blob/main/docs/maintainers/crate_release.md#regular-release

**Comment by stefano-garzarella** [comment]
> > I guess it makes sense to turn this into a workspace, but how do we handle the tags for releases?
> > Now we only have vhost tags, should we add vhost-user-backend tags as well?
> > Should we use a prefix? (e.g. vhost-user-backend-v0.7.0). Or should we align the two versions and use a single tag?
> 
> This is something that we had to do for vm-virtio as well. Initially the crates were in separate repositories. For releases we're just using the name of the crate in the tags. There are some other things that are different with workspaces when doing releases, and they're documented here: https://github.com/rust-vmm/community/blob/main/docs/maintainers/crate_release.md#regular-release

Thanks for the pointer!

So I think we should update the tags after merging this PR, right?

**Comment by andreeaflorescu** [comment]
> So I think we should update the tags after merging this PR, right?

Yes, I think that would be the right approach.

**Comment by stefano-garzarella** [comment]
@vireshk I would add a README.md in the root to describe briefly the workspace and the crates provided.
The rest LTGM, thanks for this effort!

**Review by stefano-garzarella — APPROVED** [review]
Just a minor thing if you want to re-spin, but I'm happy with the current status!

**Comment by stefano-garzarella** [comment]
> I have added minimal README now. Thanks.

Great, thanks!

> 
> [FWIW, I am on holidays (2 weeks) at the moment, though still have access to my desktop for a day or two more.]

Enjoy your holidays ;-)

File: README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# vhost
+
+The `vhost` workspace hosts libraries related to the vhost and vhost-user
+protocols.  It currently consists of the following crates:
+
+- `vhost` -> A pure rust library for vDPA, vhost and vhost-user.
+- `vhost-user-backend` -> It provides a framework to implement `vhost-user`
+  backend services,
```
[the remark this answers is not in the document]
Comment (vireshk) [reply README.md]: Fixed that now.

**Review by stefano-garzarella — APPROVED** [review]

**Review by jiangliu — APPROVED** [review]

**Comment by vireshk** [comment]
Thanks @jiangliu 

I now see that @sboeuf has added few more patches to vhost yesterday. Should I generate this tree again ? Or is there an easier way to get this sorted ?

@andreeaflorescu

**Review by jiangliu — APPROVED** [review]

**Review by stefano-garzarella — APPROVED** [review]

**Comment by vireshk** [comment]
@stefano-garzarella will you be merging it as well ?

**Comment by stefano-garzarella** [comment]
> @stefano-garzarella will you be merging it as well ?
> 
> 

@vireshk I'm not able since there are CI failures. @andreeaflorescu how can we force this?

**Comment by lauralt** [comment]
> > @stefano-garzarella will you be merging it as well ?
> 
> @vireshk I'm not able since there are CI failures. @andreeaflorescu how can we force this?

Hi, Andreea is on vacation and will be back on Friday. Is it possible to fix the commit message or do we have to merge it with admin rights?

**Comment by vireshk** [comment]
> Hi, Andreea is on vacation and will be back on Friday. Is it possible to fix the commit message or do we have to merge it with admin rights?

With admin rights. We don't want to change the commit history of the vhost-user-backend crate and so the commit stays as is.

**Comment by stefano-garzarella** [comment]
> > Hi, Andreea is on vacation and will be back on Friday. Is it possible to fix the commit message or do we have to merge it with admin rights?
> 
> With admin rights. We don't want to change the commit history of the vhost-user-backend crate and so the commit stays as is.

Yep, I agree! I think we can wait when Andreea comes back ;-)

**Comment by andreeaflorescu** [comment]
Hmmm, it does not work with admin rights either because it cannot be rebased. Are you editing the history of the crate? The commits should all be added on top, and thus there shouldn't be a problem.

**Comment by vireshk** [comment]
> Hmmm, it does not work with admin rights either because it cannot be rebased. Are you editing the history of the crate? The commits should all be added on top, and thus there shouldn't be a problem.

I am simply doing a git merge here. The patches from vhost-user-backend are new to vhost crate and so the errors, as they are treated as new.

**Comment by andreeaflorescu** [comment]
@vireshk could you try to run the commands suggested in this stack overflow post? https://stackoverflow.com/questions/50012036/github-cant-rebase-my-feature-branch-this-branch-cannot-be-rebased-due-to-con

**Comment by vireshk** [comment]
> @vireshk could you try to run the commands suggested in this stack overflow post? https://stackoverflow.com/questions/50012036/github-cant-rebase-my-feature-branch-this-branch-cannot-be-rebased-due-to-con

Which commands exactly from there ?

All I see is rebase and that will change the history of vhost-user-backend. IIUC, it suggests to checkout vhost-user-backend branch and then rebase over main (from vhost repository). That will change the commit-ids of all patches from vhost-user-backend. Moreover, since we need to move all of them to a different path, under crates/ i.e., we will have conflicts, which I have currently solved in the Merge itself.

**Comment by vireshk** [comment]
@andreeaflorescu As an admin, can't you go to:

Settings -> Code and Automation -> Branches -> Branch Protection Rule 
and Disable "Require status checks to pass before merging" checkbox in the rule.

Merge the branch.

And then reenable the check ? It should be doable I think.

**Comment by andreeaflorescu** [comment]
I did disable the required checks, the PR still cannot be merged because as I said in my previous comment the commits in this PR cannot be rebased on top of the existing commits in the repository. That's why I was suggesting to run the following commands to fix these (from the stack overflow post):

```
$ git rebase origin/master -i          # 1
$ git push -f origin/workspace # 2
```

This is the only thing blocking the merge. All the other changes can be bypassed. We do not have signed commits as a requirement for merges.

**Comment by andreeaflorescu** [comment]
LE: chatted offline with @vireshk. This PR cannot be merged with rebase, it must have a merge commit. I temporarily allowed merge commits on the repository to be able to merge this.

**Pull request merged**